File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> keymap, vector<string> targets) {
    vector<int> answer;
    
    int max_val = 1e9;
    vector<int> dp(26, max_val);
    
    for(auto& cur_s : keymap){
        
        int cnt = 1;
        for(int i = 0 ; i < cur_s.size() ; i++){
            int cur = (cur_s[i] - 'A');
            dp[cur] = min(dp[cur], cnt);
            cnt++;
        }
        
        
    }
    
    
    for(auto& cur_s : targets){
        int val = 0;
        for(auto& c : cur_s){
            int cur = c - 'A';
            if(dp[cur] == max_val){
                val = -1;
                break;
            }else{
                val += dp[cur];
            }
        }
        
        answer.push_back(val);
    }
    
    
    
    return answer;
}

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.*;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        List<Integer> list = new ArrayList<>();
        
        for (String t : targets) {
            int val = 0;
            for (char cur : t.toCharArray()) {
                int min = Integer.MAX_VALUE;
                for (String key : keymap) {
                    int index = key.indexOf(cur);
                    if (index != -1) {
                        min = Math.min(min, index + 1);
                    }
                }
                if (min == Integer.MAX_VALUE) {
                    val = -1;
                    break;
                }
                val += min;
            }
            list.add(val);
        }
        
        return list.stream().mapToInt(Integer::intValue).toArray();
    }
}